vtl-spark/src/main/java/fr/insee/vtl/spark/SparkDataset.java

```diff
@@ -80,39 +68,77 @@ public SparkDataset(Dataset vtlDataset, Map<String, Role> roles, SparkSession sp
         vtlDataset.getDataPoints().stream()
             .map(points -> RowFactory.create(points.toArray(new Object[] {})))
             .collect(Collectors.toList());
-
     // TODO: Handle nullable with component
     StructType schema = toSparkSchema(vtlDataset.getDataStructure());
-
     this.sparkDataset = spark.createDataFrame(rows, schema);
     this.roles = Objects.requireNonNull(roles);
   }
 
-  /** Cast integer and float types to long and double. */
+  /**
+   * Constructor taking a Spark dataset and a mapping of component names and roles.
+   *
+   * @param sparkDataset a Spark dataset.
+   * @param roles a map between component names and their roles in the dataset.
+   */
+  public SparkDataset(org.apache.spark.sql.Dataset<Row> sparkDataset, Map<String, Role> roles) {
+    org.apache.spark.sql.Dataset<Row> castedSparkDataset =
+        castIfNeeded(Objects.requireNonNull(sparkDataset));
+    DataStructure dataStructure = fromSparkSchema(sparkDataset.schema(), roles, Map.of());
+    this.sparkDataset = addMetadata(castedSparkDataset, dataStructure);
+    this.roles = Objects.requireNonNull(roles);
+  }
+
+  /** Cast integer and float types to long and double efficiently. */
   private static org.apache.spark.sql.Dataset<Row> castIfNeeded(
       org.apache.spark.sql.Dataset<Row> sparkDataset) {
-    var casted = sparkDataset;
     StructType schema = sparkDataset.schema();
-    for (StructField field : JavaConverters.asJavaCollection(schema)) {
-      if (IntegerType.sameType(field.dataType())) {
-        casted = casted.withColumn(field.name(), casted.col(field.name()).cast(LongType));
-      } else if (FloatType.sameType(field.dataType())) {
-        casted = casted.withColumn(field.name(), casted.col(field.name()).cast(DoubleType));
-      } else if (DecimalType.class.equals(field.dataType().getClass())) {
-        casted = casted.withColumn(field.name(), casted.col(field.name()).cast(DoubleType));
-      }
-    }
-    return casted;
+
+    // Se construye una lista de expresiones para castear en una sola transformación
+    List<Column> castedColumns =
+        Arrays.stream(schema.fields())
+            .map(
+                field -> {
+                  DataType type = field.dataType();
+                  Column col = sparkDataset.col(field.name());
+                  if (type instanceof IntegerType
+                      || type instanceof FloatType
+                      || type instanceof DecimalType) {
+                    return col.cast(
+                            type instanceof IntegerType ? DataTypes.LongType : DataTypes.DoubleType)
+                        .alias(field.name());
+                  }
+                  return col;
+                })
+            .collect(Collectors.toList());
+
+    return sparkDataset.select(castedColumns.toArray(new Column[0]));
```

Comment:
Consider `toArray();`
```suggestion
    var castColumns =
        Arrays.stream(schema.fields())
            .map(
                field -> {
                  DataType type = field.dataType();
                  Column col = sparkDataset.col(field.name());
                  if (type instanceof IntegerType
                      || type instanceof FloatType
                      || type instanceof DecimalType) {
                    return col.cast(
                            type instanceof IntegerType ? DataTypes.LongType : DataTypes.DoubleType)
                        .alias(field.name());
                  }
                  return col;
                })
            .toArray(Column[]::new));

    return sparkDataset.select(castColumns);
```